sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ResolveDefaultColumns.scala

```diff
@@ -18,14 +18,227 @@
+ * @param catalog the catalog to use for looking up the schema of INSERT INTO table objects.
+ * @param insert the enclosing INSERT statement for which this rule is processing the query, if any.
+ */
+case class ResolveDefaultColumns(
+  analyzer: Analyzer,
+  catalog: SessionCatalog,
+  insert: Option[InsertIntoStatement] = None) extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsWithPruning(
+    (_ => SQLConf.get.enableDefaultColumns), ruleId) {
+    case i@InsertIntoStatement(_, _, _, _, _, _)
+      if i.query.collectFirst { case u: UnresolvedInlineTable => u }.isDefined =>
+      // Create a helper instance of this same rule with the `insert` argument populated as `i`.
+      // Then recursively apply it on the `query` of `i` to transform the result. This recursive
+      // application lets the below case matching against `UnresolvedInlineTable` to trigger
+      // anywhere that operator may appear in the descendants of `i.query`.
+      val helper = ResolveDefaultColumns(analyzer, catalog, Some(i))
+      val newQuery = helper.apply(i.query)
+      i.copy(query = newQuery)
+
+    case table: UnresolvedInlineTable
+      if insert.isDefined &&
+        table.rows.nonEmpty && table.rows.forall(_.size == table.rows(0).size) =>
+      val expanded: UnresolvedInlineTable = addMissingDefaultColumnValues(table).getOrElse(table)
+      replaceExplicitDefaultColumnValues(analyzer, expanded).getOrElse(table)
+
+    case i@InsertIntoStatement(_, _, _, project: Project, _, _)
+      if !project.resolved =>
+      val helper = ResolveDefaultColumns(analyzer, catalog, Some(i))
+      val replaced: Option[Project] = helper.replaceExplicitDefaultColumnValues(analyzer, project)
+      if (replaced.isDefined) i.copy(query = replaced.get) else i
+
+    case i@InsertIntoStatement(_, _, _, project: Project, _, _)
+      if project.resolved =>
+      val helper = ResolveDefaultColumns(analyzer, catalog, Some(i))
+      val expanded: Project = helper.addMissingDefaultColumnValues(project).getOrElse(project)
+      val replaced: Option[Project] = helper.replaceExplicitDefaultColumnValues(analyzer, expanded)
+      if (replaced.isDefined) i.copy(query = replaced.get) else i
+  }
+
+  // Helper method to check if an expression is an explicit DEFAULT column reference.
+  private def isExplicitDefaultColumn(expr: Expression): Boolean = expr match {
+    case u: UnresolvedAttribute if u.name.equalsIgnoreCase(CURRENT_DEFAULT_COLUMN_NAME) => true
+    case _ => false
+  }
+
+  // Each of the following methods adds a projection over the input plan to generate missing default
+  // column values.
+  private def addMissingDefaultColumnValues(
+      table: UnresolvedInlineTable): Option[UnresolvedInlineTable] = {
+    assert(insert.isDefined)
+    val numQueryOutputs: Int = table.rows(0).size
+    val schema: StructType = getInsertTableSchema.getOrElse(return None)
+    val schemaWithoutPartitionCols =
```

Comment:
It seems that we always need `schemaWithoutPartitionCols` instead of `schema`. Shall we change the method `getInsertTableSchema` to `getInsertTableSchemaWithoutPartitonCols`?

Reply:
Good idea! This is much simpler now, done. I updated the method name to help remind callers that the partition columns are dropped.